core/controller.go

```diff
@@ -268,6 +303,51 @@ func (c *Controller) StopStreamHandler(w http.ResponseWriter, r *http.Request, p
 	w.WriteHeader(http.StatusOK)
 }
 
+func (c *Controller) StartPreloadStream(Name string, Uri string) {
+	var dto StreamDTO
+
+	dto.URI = Uri
+	logrus.Debugf("%s is being initialized", dto.URI)
+
+	_, knownStream := c.index[dto.URI]
+	if knownStream {
+		return
+	}
+
+	stream, id := streamer.NewStream(
+		dto.URI,
+		c.spec.StoreDir,
+		c.spec.KeepFiles,
+		c.spec.Audio,
+		streamer.ProcessLoggingOpts{
+			Enabled:    c.spec.ProcessLogging.Enabled,
+			Compress:   c.spec.ProcessLogging.Compress,
+			Directory:  c.spec.ProcessLogging.Directory,
+			MaxAge:     c.spec.ProcessLogging.MaxAge,
+			MaxBackups: c.spec.ProcessLogging.MaxBackups,
+			MaxSize:    c.spec.ProcessLogging.MaxSize,
+		},
+		25*time.Second,
+	)
+
+	streamname := id
+	stream.Start().Wait()
+	if stream.Running {
```

Comment:
We can avoid this being in a bigger if else by reversing the question. Also we can think about how we want to deal with preloaded streams not wanting to start properly.

What do you think about something like this?
```go
	if !stream.Running {
        if c.blacklist.AddOrIncrease(URI).IsBanned(URI) {
           delete(c.preload, Name)
        }
        return
	}
    c.streams[id] = stream
	c.index[URI] = id
	if len(Name) > 0 {
		c.alias[Name] = id
		streamname = Name
	}
	c.blacklist.Remove(URI)
	delete(c.preload, Name)
	logrus.Infoln("started stream /stream/" + streamname + "/index.m3u8")
```

If it is banned according to the blacklist then it is getting removed, if not it still provides the opportunity to try it a couple of times.

Reply:
Nice! I'll make that change